src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,188 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    public MyLinkedList() {
+        this.head = null;
+        this.tail = null;
+        this.size = 0;
+    }
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(value, null, tail);
+        if (tail == null) {
+            head = newNode;
+            tail = newNode;
+        } else {
+            tail.setNext(newNode);
+            newNode.setPrev(tail);
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index out of bounds");
+        }
+        if (index == size) {
+            add(value);
+            return;
+        }
+        Node<T> newNode;
+        if (index == 0) {
+            newNode = new Node<>(value, head, null);
+            head.setPrev(newNode);
+            head = newNode;
+        } else {
+            Node<T> current = getNode(index);
+            newNode = new Node<>(value, current, current.getPrev());
+            current.getPrev().setNext(newNode);
+            current.setPrev(newNode);
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return getNode(index).getData();
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = getNode(index);
+        T oldValue = node.getData();
+        node.setData(value);
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> nodeToRemove = getNode(index);
+        if (nodeToRemove.getPrev() != null) {
+            nodeToRemove.getPrev().setNext(nodeToRemove.getNext());
+        } else {
+            head = nodeToRemove.getNext();
+        }
+        if (nodeToRemove.getNext() != null) {
+            nodeToRemove.getNext().setPrev(nodeToRemove.getPrev());
+        } else {
+            tail = nodeToRemove.getPrev();
+        }
+        size--;
+        return nodeToRemove.getData();
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        while (current != null) {
+            if ((current.getData() == object)
+                    || (current.getData() != null && current.getData().equals(object))) {
```

Comment:
The condition for checking if the current node's data equals the object should handle null values separately. Consider adding a separate check for `object == null` to avoid calling `equals` on a null reference.